### PR TITLE
Link company with dnb admin step 2

### PR DESCRIPTION
**Pull request by currycoder**
### Description of change
A view was added to allow admin users to review and confirm changes that will be made for D&B company linking.

I'd recommend reviewing the commits in order.

### Checklist

* [X] Has a new newsfragment been created? Check [changelog/README.md](https://github.com/uktrade/data-hub-api/blob/master/changelog/README.md) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-api/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/dnb_api/tasks.py [hunk 1]
```diff
@@ -25,17 +25,26 @@
...
     except DNBServiceError as exc:
-        if is_server_error(exc.status_code):
+        if is_server_error(exc.status_code) and retry_failures:
             raise task.retry(exc=exc, countdown=60)
         raise
     except (DNBServiceConnectionError, DNBServiceTimeoutError) as exc:
-        raise task.retry(exc=exc, countdown=60)
+        if retry_failures:
+            raise task.retry(exc=exc, countdown=60)
+        else:
```
Comment (reupen) [inline datahub/dnb_api/tasks.py]: Should be able to remove this `else:`.

File: datahub/company/admin/dnb_link/review_changes.py [hunk 2]
```diff
@@ -1,15 +1,122 @@
...
+        link_company_with_dnb(dh_company_id, duns_number, user)
+    except serializers.ValidationError:
+        message = 'Data from D&B did not pass the Data Hub validation checks.'
+        raise AdminException(message, error_url)
+    except (
+        DNBServiceError,
+        DNBServiceConnectionError,
+        DNBServiceTimeoutError,
+        DNBServiceInvalidResponse,
+    ):
```
Comment (reupen) [inline datahub/company/admin/dnb_link/review_changes.py]: I think we could use an exception hierarchy to avoid having to handle four exceptions here (but might be something for another PR).

File: datahub/company/admin/dnb_link/review_changes.py [hunk 3]
```diff
@@ -1,15 +1,122 @@
...
     )
-    return HttpResponseRedirect(company_change_url)
+
+    form = SelectIdsToLinkForm(data=request.GET)
+    if not form.is_valid():
+        message = _build_error_message(form.errors)
+        raise AdminException(message, company_list_page)
+
+    dh_company = form.cleaned_data['company']
+    duns_number = request.GET.get('duns_number', form.cleaned_data['duns_number'])
```
Comment (alixedi) [inline datahub/company/admin/dnb_link/review_changes.py]: I presume we are doing this as opposed to:

```
duns_number = form.cleaned_data['duns_number']
```

to enable a direct link to the "review changes" page? If so, it would be nice to include a comment here.

File: datahub/company/admin/dnb_link/review_changes.py [hunk 4]
```diff
@@ -1,15 +1,122 @@
...
+    get_company,
+)
 
+
+def _build_error_message(all_errors):
+    error_message = ''
+    for field, errors in all_errors.items():
+        concatenated_errors = ', '.join(errors)
+        error_message += f'{field}: {concatenated_errors} '
+    return error_message
```
Comment (reupen) [inline datahub/company/admin/dnb_link/review_changes.py]: It might be simpler to flatten the list of errors completely e.g.:

```python

messages = [
    f'{field}: {error}'
    for field, errors in all_errors.items()
    for error in errors
]

return ', '.join(messages)  # or maybe newline delimited
```

But, I also think there might be too much display formatting here. I think it would be even better to be able to attach a list of error messages to the exception and have the exception handler deal with them (for example, it might call `messages.add_message()` multiple times).

File: datahub/company/admin/dnb_link/review_changes.py (hunk 3 above)
Comment (reupen) [reply datahub/company/admin/dnb_link/review_changes.py]: Yes, it would be good to understand this, as `request.GET` is fed into the form.

File: datahub/company/admin/dnb_link/review_changes.py [hunk 5]
```diff
@@ -1,15 +1,122 @@
...
+@method_decorator(csrf_protect)
 def dnb_link_review_changes(model_admin, request):
     """
     View to allow users to review changes that would be applied to a record before linking it.
     POSTs make the link and redirect the user to view the updated record.
     """
-    # TODO: Remove this temporary redirect
-    company_change_url = reverse(
+    if not model_admin.has_change_permission(request):
+        raise PermissionDenied
```
Comment (reupen) [inline datahub/company/admin/dnb_link/review_changes.py]:
```suggestion
        raise PermissionDenied()
```

File: datahub/company/admin/dnb_link/review_changes.py (hunk 3 above)
Comment (currycoder) [reply datahub/company/admin/dnb_link/review_changes.py]: My bad!  This is a leftover after refactoring from a point where this view didn't use a form - Ali is right, it should be:
`duns_number = form.cleaned_data['duns_number']`

File: datahub/company/admin/dnb_link/review_changes.py (hunk 2 above)
Comment (currycoder) [reply datahub/company/admin/dnb_link/review_changes.py]: By exception hierarchy, do you mean ensuring that these all inherit from a common base class?

File: datahub/company/admin/dnb_link/review_changes.py (hunk 4 above)
Comment (currycoder) [reply datahub/company/admin/dnb_link/review_changes.py]: I'll have a think about this.  Could this perhaps wait for a future cleanup PR?

Comment (currycoder) [reply datahub/company/admin/dnb_link/review_changes.py]: I do like the idea of calling `messages.add_message()` multiple times.

File: datahub/company/admin/dnb_link/review_changes.py (hunk 2 above)
Comment (reupen) [reply datahub/company/admin/dnb_link/review_changes.py]: Yep, that's what I meant.

File: datahub/company/admin/dnb_link/review_changes.py (hunk 4 above)
Comment (reupen) [reply datahub/company/admin/dnb_link/review_changes.py]: > I'll have a think about this. Could this perhaps wait for a future cleanup PR?
Yep :+1:

File: datahub/company/admin/dnb_link/review_changes.py (hunk 2 above)
Comment (currycoder) [reply datahub/company/admin/dnb_link/review_changes.py]: Ok neat - I've added this to a cleanup ticket which I'll get going with shortly after the feature gets released.
